### PR TITLE
Add `process_slashings`

### LINKED ISSUE #322: Implement beacon chain state transition `process_slashings`
### What is wrong?

The spec declares functionality to process slashings every epoch (across the entire validator set, rather than at the level of the individual validator). This functionality is missing right now.

Reference: https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#validator-registry-and-shuffling-seed-data, scroll down to Python block with function `process_slashings`.

### How can it be fixed

Implement the spec :)

edit: there is a `TODO` comment for the function in #324.

### PR DESCRIPTION
**Pull request by hwwhww**
### What was wrong?
Fix #322

### How was it fixed?

Implement as the spec.

#### Cute Animal Picture

puma-4029263_640

### CODE REVIEW

File: tests/eth2/beacon/state_machines/forks/test_serenity_epoch_processing.py [hunk 1]
```diff
@@ -1063,6 +1066,147 @@ def mock_generate_seed(state,
...
+                                    latest_slashed_exit_length,
+                                    total_penalties,
+                                    total_balance,
+                                    expected_penalty):
+    state = genesis_state.copy(
+        slot=get_epoch_start_slot(current_epoch, slots_per_epoch),
+    )
+    validator_index = 0
+    total_penalties = 10 ** 9  # 1 ETH
+    penalty = penalty = _compute_individual_penalty(
```
Comment (NIC619) [inline tests/eth2/beacon/state_machines/forks/test_serenity_epoch_processing.py]:
```suggestion
    penalty = _compute_individual_penalty(
```

File: tests/eth2/beacon/state_machines/forks/test_serenity_epoch_processing.py [hunk 2]
```diff
@@ -1063,6 +1066,147 @@ def mock_generate_seed(state,
...
+                                    current_epoch,
+                                    latest_slashed_exit_length,
+                                    total_penalties,
+                                    total_balance,
+                                    expected_penalty):
+    state = genesis_state.copy(
+        slot=get_epoch_start_slot(current_epoch, slots_per_epoch),
+    )
+    validator_index = 0
+    total_penalties = 10 ** 9  # 1 ETH
```
Comment (NIC619) [inline tests/eth2/beacon/state_machines/forks/test_serenity_epoch_processing.py]: Redefined `total_penalties`

File: eth2/beacon/state_machines/forks/serenity/epoch_processing.py [hunk 3]
```diff
@@ -1020,6 +1020,78 @@ def _update_latest_active_index_roots(state: BeaconState,
...
+            penalty = _compute_individual_penalty(
+                state=state,
+                config=config,
+                validator_index=validator_index,
+                total_penalties=total_penalties,
+                total_balance=total_balance,
+            )
+            state = state.update_validator_balance(
+                validator_index=validator_index,
+                balance=state.validator_balances[validator_index] - penalty,
```
Comment (NIC619) [inline eth2/beacon/state_machines/forks/serenity/epoch_processing.py]: Should we prevent balance underflow here?

File: tests/eth2/beacon/state_machines/forks/test_serenity_epoch_processing.py [hunk 4]
```diff
@@ -1063,6 +1066,147 @@ def mock_generate_seed(state,
...
+        latest_slashed_balances=latest_slashed_balances,
+    )
+    validator_index = 0
+    validator = state.validator_registry[validator_index].copy(
+        slashed=True,
+        withdrawable_epoch=current_epoch + latest_slashed_exit_length // 2
+    )
+    state = state.update_validator_registry(validator_index, validator)
+
+    slashing_validator_index = 0
```
Comment (NIC619) [inline tests/eth2/beacon/state_machines/forks/test_serenity_epoch_processing.py]: `slashing_validator_index` seems to be a duplicate of `validator_index`

File: eth2/beacon/state_machines/forks/serenity/epoch_processing.py (hunk 3 above)
Comment (hwwhww) [reply eth2/beacon/state_machines/forks/serenity/epoch_processing.py]: Hmm, logically, they would be ejected before it's under the risk of underflow.
Raise a question in the spec repo: https://github.com/ethereum/eth2.0-specs/issues/727

**Pull request merged**